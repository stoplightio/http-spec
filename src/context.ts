import { hasRef, isLocalRef } from '@stoplight/json';
// @ts-ignore
import * as fnv from 'fnv-plus';

import type {
  AvailableContext,
  Fragment,
  IdGenerator,
  RefResolver,
  TransformerContext,
  TranslateFunction,
} from './types';

export const DEFAULT_ID_GENERATOR: IdGenerator = input => fnv.fast1a32hex(input);

export function createContext<T extends Record<string, unknown>>(
  document: T,
  resolveRef: RefResolver<T>,
  generateId: IdGenerator,
): TransformerContext<T> {
  let context: AvailableContext = 'service';
  return {
    document,
    get context() {
      return context;
    },
    set context(value) {
      context = value;
      if (value !== 'operation') {
        this.parentId = this.ids[value];
      }
    },
    maybeResolveLocalRef(target: unknown) {
      if (hasRef(target) && isLocalRef(target.$ref)) {
        try {
          return resolveRef.call(this, target);
        } catch {
          return target;
        }
      }

      return target;
    },
    generateId(value) {
      this.parentId = generateId(value);
      return this.parentId;
    },
    ids: {
      service: '',
      path: '',
      operation: '',
    },
    parentId: '',
  };
}

export function withContext<F extends TranslateFunction<any, any[]> = TranslateFunction<Fragment, unknown[]>>(
  fn: F,
): F {
  return <F>function (...args: Parameters<F>) {
    const { context, parentId } = this;
    const result = fn.apply(this, args);
    this.context = context;
    this.parentId = parentId;
    return result;
  };
}
