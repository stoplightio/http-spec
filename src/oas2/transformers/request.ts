import { IHttpOperationRequest } from '@stoplight/types';
import { BodyParameter, FormDataParameter, Parameter } from 'swagger-schema-official';

import {
  translateFromFormDataParameters,
  translateToBodyParameter,
  translateToHeaderParam,
  translateToPathParameter,
  translateToQueryParameter,
} from './params';

export function translateToRequest(parameters: Parameter[], consumes: string[]): IHttpOperationRequest {
  const bodyParameters = parameters.filter((p): p is BodyParameter => p.in === 'body');
  const formDataParameters = parameters.filter((p): p is FormDataParameter => p.in === 'formData');
  const request: IHttpOperationRequest = {};

  // if 'body' and 'form data' defined prefer 'body'
  if (!!bodyParameters.length) {
    // There can be only one body parameter (taking first one)
    request.body = translateToBodyParameter(bodyParameters[0], consumes);
  } else if (!!formDataParameters.length) {
    request.body = translateFromFormDataParameters(formDataParameters, consumes);
  }

  return parameters.reduce(reduceRemainingParameters, request);
}

function reduceRemainingParameters(request: IHttpOperationRequest, parameter: Parameter) {
  if (parameter.in === 'query') {
    const queryParameter = translateToQueryParameter(parameter);
    request.query = (request.query || []).concat(queryParameter);
  } else if (parameter.in === 'path') {
    const pathParameter = translateToPathParameter(parameter);
    request.path = (request.path || []).concat(pathParameter);
  } else if (parameter.in === 'header') {
    const headerParameter = translateToHeaderParam(parameter);
    request.headers = (request.headers || []).concat(headerParameter);
  }
  return request;
}
