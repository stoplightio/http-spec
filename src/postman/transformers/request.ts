import {
  HttpParamStyles,
  IHttpHeaderParam,
  IHttpOperationRequest,
  IHttpOperationRequestBody,
  IHttpPathParam,
  IHttpQueryParam,
  IMediaTypeContent,
} from '@stoplight/types';
import { JSONSchema4 } from 'json-schema';
import {
  FormParam,
  HeaderDefinition,
  HeaderList,
  PropertyList,
  QueryParam,
  Request,
  RequestBody,
} from 'postman-collection';
import * as toJsonSchema from 'to-json-schema';
import { transformDescriptionDefinition, transformValueToHttpParam } from '../util';

export function transformRequest(request: Request): IHttpOperationRequest {
  return {
    query: request.url.query.all().map(transformQueryParam),
    headers: request.headers.all().map(transformHeader),
    path: transformPathParams(request.url.path),
    body: request.body ? transformBody(request.body, findContentType(request.headers)) : undefined,
  };
}

function findContentType(headers: HeaderList) {
  const header = headers.all().find(h => h.key.toLowerCase() === 'content-type');
  return header ? header.value.toLowerCase() : undefined;
}

export function transformQueryParam(queryParam: QueryParam): IHttpQueryParam {
  return {
    name: queryParam.key as string, // no key no game
    style: HttpParamStyles.Form,
    ...(queryParam.value ? transformValueToHttpParam(queryParam.value) : undefined),
  };
}

export function transformHeader(header: HeaderDefinition): IHttpHeaderParam {
  return {
    name: header.key as string,
    style: HttpParamStyles.Simple,
    ...(header.value ? transformValueToHttpParam(header.value) : undefined),
  };
}

export function transformPathParams(segments: string[]): IHttpPathParam[] {
  return segments.reduce<IHttpHeaderParam[]>((params, segment) => {
    if (segment.startsWith(':')) {
      params.push({
        name: segment.substring(1),
        style: HttpParamStyles.Simple,
      });
    }

    return params;
  }, []);
}

export function transformBody(body: RequestBody, mediaType?: string): IHttpOperationRequestBody | undefined {
  switch (body.mode) {
    case 'raw':
      if (!body.raw) return;
      return { contents: [transformRawBody(body.raw, mediaType)] };

    case 'formdata':
      if (!body.formdata) return;
      return {
        contents: [transformParamsBody<FormParam>(body.formdata, mediaType || 'multipart/form-data')],
      };

    case 'urlencoded':
      if (!body.urlencoded) return;
      return {
        contents: [transformParamsBody<QueryParam>(body.urlencoded, mediaType || 'application/x-www-form-urlencoded')],
      };
  }

  return;
}

function transformRawBody(raw: string, mediaType?: string): IMediaTypeContent {
  try {
    const parsed = JSON.parse(raw);

    return {
      mediaType: mediaType || 'application/json',
      examples: [
        {
          key: 'default',
          value: parsed,
        },
      ],
      schema: toJsonSchema(parsed) as JSONSchema4,
    };
  } catch (e) {
    /* noop */
  }

  return {
    mediaType: mediaType || 'text/plain',
    examples: [
      {
        key: 'default',
        value: raw,
      },
    ],
  };
}

function transformParamsBody<T extends FormParam | QueryParam>(
  params: PropertyList<T>,
  mediaType: string,
): IMediaTypeContent {
  return {
    mediaType,
    schema: {
      type: 'object',
      properties: params.all().reduce<NonNullable<JSONSchema4['properties']>>((props, item) => {
        // @todo: !
        props[item.key!] = {
          type: 'string',
          description: item.description && transformDescriptionDefinition(item.description),
        };
        return props;
      }, {}),
    },
    examples: [
      {
        key: 'default',
        value: params.all().reduce((value, item) => {
          // @todo: !
          value[item.key!] = item.value;
          return value;
        }, {}),
      },
    ],
  };
}
