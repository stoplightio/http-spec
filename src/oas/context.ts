// @ts-expect-error: no types
import * as fnv from 'fnv-plus';

import { createContext as _createContext } from '../context';
import type { Fragment, IdGenerator } from '../types';
import { resolveRef } from './resolver';

const HASH_32_KINDS = ['http_operation', 'schema'];

function getKind(input: string) {
  const i = input.indexOf('-');
  return input.slice(0, i);
}

const DEFAULT_ID_GENERATOR: IdGenerator = input => {
  const kind = getKind(input);
  return HASH_32_KINDS.some(k => kind === k) ? fnv.fast1a32hex(input) : fnv.fast1a52hex(input);
};

export function createContext<T extends Fragment>(document: T) {
  return _createContext<T>(document, resolveRef, DEFAULT_ID_GENERATOR);
}
