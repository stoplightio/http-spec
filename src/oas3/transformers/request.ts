import { isPlainObject } from '@stoplight/json';
import type {
  IHttpHeaderParam,
  IHttpOperationRequest,
  IHttpOperationRequestBody,
  IHttpParam,
  INodeExample,
  INodeExternalExample,
  Optional,
} from '@stoplight/types';
import { HttpParamStyles } from '@stoplight/types';
import type { JSONSchema7 } from 'json-schema';
import type { ParameterObject } from 'openapi3-ts';
import pickBy = require('lodash.pickby');

import { isNonNullable, isString } from '../../guards';
import { OasVersion } from '../../oas';
import { getValidOasParameters } from '../../oas/accessors';
import { isValidParamStyle } from '../../oas/guards';
import { translateSchemaObject } from '../../oas/transformers/schema';
import type { ArrayCallbackParameters, Fragment } from '../../types';
import { entries } from '../../utils';
import { isRequestBodyObject } from '../guards';
import type { Oas3TranslateFunction } from '../types';
import { translateMediaTypeObject } from './content';

export const translateRequestBody: Oas3TranslateFunction<[requestBodyObject: unknown], IHttpOperationRequestBody> =
  function (requestBodyObject) {
    const resolvedRequestBodyObject = this.maybeResolveLocalRef(requestBodyObject);

    if (isRequestBodyObject(resolvedRequestBodyObject)) {
      return {
        contents: entries(resolvedRequestBodyObject.content).map(translateMediaTypeObject, this).filter(isNonNullable),

        ...pickBy(
          {
            required: resolvedRequestBodyObject.required,
            description: resolvedRequestBodyObject.description,
          },
          isNonNullable,
        ),
      };
    }

    return { contents: [] };
  };

const translateParameterObjectSchema: Oas3TranslateFunction<
  [parameterObject: Fragment],
  Optional<JSONSchema7>
> = function (parameterObject) {
  if (!isPlainObject(parameterObject.schema)) return;

  return translateSchemaObject.call(this, {
    ...parameterObject.schema,
    ...('example' in parameterObject ? { example: parameterObject.example } : null),
  });
};

const translateToExample: Oas3TranslateFunction<
  ArrayCallbackParameters<[key: string, example: unknown]>,
  Optional<INodeExample | INodeExternalExample>
> = function ([key, example]) {
  if (!isPlainObject(example)) return;

  if (!('value' in example) && typeof example.externalValue !== 'string') return;

  return {
    key,

    ...(typeof example.externalValue === 'string'
      ? { externalValue: example.externalValue }
      : { value: example.value }),

    ...pickBy(
      {
        summary: example.summary,
        description: example.description,
      },
      isString,
    ),
  };
};

export const translateParameterObject: Oas3TranslateFunction<[parameterObject: ParameterObject], IHttpParam> =
  function (parameterObject) {
    const examples = entries(parameterObject.examples).map(translateToExample, this).filter(isNonNullable);

    const hasDefaultExample = examples.some(({ key }) => key.includes('default'));
    const schema = translateParameterObjectSchema.call(this, parameterObject);

    return {
      name: parameterObject.name,
      deprecated: !!parameterObject.deprecated,
      style: isValidParamStyle(parameterObject.style) ? parameterObject.style : HttpParamStyles.Simple,
      explode: !!(parameterObject.explode ?? parameterObject.style === HttpParamStyles.Form),
      examples:
        'example' in parameterObject && !hasDefaultExample
          ? [{ key: 'default', value: parameterObject.example }, ...examples]
          : examples,

      ...pickBy(
        {
          description: parameterObject.description,
        },
        isString,
      ),

      ...pickBy(
        {
          schema,
          content: parameterObject.content,
        },
        isPlainObject,
      ),
    };
  };

export const translateToRequest: Oas3TranslateFunction<[path: Fragment, operation: Fragment], IHttpOperationRequest> =
  function (path, operation) {
    const params: Omit<IHttpOperationRequest, 'header'> & { header: IHttpHeaderParam[] } = {
      header: [],
      query: [],
      cookie: [],
      path: [],
    };

    for (const param of getValidOasParameters(this.document, OasVersion.OAS3, operation.parameters, path.parameters)) {
      const { in: key } = param;
      const target = params[key];
      if (!Array.isArray(target)) continue;

      target.push(translateParameterObject.call(this, param) as any);
    }

    return {
      body: translateRequestBody.call(this, operation?.requestBody),
      headers: params.header,
      query: params.query,
      cookie: params.cookie,
      path: params.path,
    };
  };
