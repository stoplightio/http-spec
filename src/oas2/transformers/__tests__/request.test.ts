import {
  BodyParameter,
  FormDataParameter,
  HeaderParameter,
  PathParameter,
  QueryParameter,
} from 'swagger-schema-official';

import { createContext, DEFAULT_ID_GENERATOR } from '../../../context';
import {
  translateFromFormDataParameters,
  translateToBodyParameter,
  translateToHeaderParam,
  translateToPathParameter,
  translateToQueryParameter,
} from '../params';
import { translateToRequest as _translateToRequest } from '../request';

jest.mock('../params');

const translateToRequest = (path: Record<string, unknown>, parameters: any[]) => {
  const ctx = createContext({ consumes: ['*'], paths: { '/api': { parameters } } }, DEFAULT_ID_GENERATOR);
  return _translateToRequest.call(ctx, path, { parameters });
};

describe('request', () => {
  const fakeParameter: FormDataParameter = {
    name: 'name',
    type: 'string',
    in: 'formData',
  };
  const fakeBodyParameter: BodyParameter = { in: 'body', name: 'param' };
  const fakeFormParameter: FormDataParameter = { in: 'formData', name: 'param', type: 'number' };
  const fakeQueryParameter: QueryParameter = { in: 'query', name: 'param' };
  const fakeHeaderParameter: HeaderParameter = { in: 'header', name: 'param' };
  const fakePathParameter: PathParameter = { in: 'path', name: 'param', required: true };

  beforeEach(() => {
    (translateToBodyParameter as jest.Mock).mockReturnValue(fakeBodyParameter);
    (translateFromFormDataParameters as jest.Mock).mockReturnValue(fakeFormParameter);
    (translateToQueryParameter as jest.Mock).mockReturnValue(fakeQueryParameter);
    (translateToPathParameter as jest.Mock).mockReturnValue(fakePathParameter);
    (translateToHeaderParam as jest.Mock).mockReturnValue(fakeHeaderParameter);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('given single body param should translate to request with body', () => {
    expect(translateToRequest({}, [fakeBodyParameter])).toMatchSnapshot();
  });

  it('given single form param should translate to request with form', () => {
    expect(translateToRequest({}, [fakeFormParameter])).toMatchSnapshot();
  });

  it('given single path param should translate to request with path', () => {
    expect(translateToRequest({}, [fakePathParameter])).toMatchSnapshot();
  });

  it('given single query param should translate to request with query', () => {
    expect(translateToRequest({}, [fakeQueryParameter])).toMatchSnapshot();
  });

  it('given single header param should translate to request with header', () => {
    expect(translateToRequest({}, [fakeHeaderParameter])).toMatchSnapshot();
  });

  it('given two query params should translate', () => {
    expect(translateToRequest({}, [fakeQueryParameter, fakeQueryParameter])).toMatchSnapshot();
  });

  it('given two header params should translate', () => {
    expect(translateToRequest({}, [fakeHeaderParameter, fakeHeaderParameter])).toMatchSnapshot();
  });

  it('given two path params should translate', () => {
    expect(translateToRequest({}, [fakePathParameter, fakePathParameter])).toMatchSnapshot();
  });

  it('should translate mixed request', () => {
    expect(
      translateToRequest({}, [fakeParameter, fakeParameter, fakeParameter, fakeParameter, fakeParameter]),
    ).toMatchSnapshot();
  });
});
