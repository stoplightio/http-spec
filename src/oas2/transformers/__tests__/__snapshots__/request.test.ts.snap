// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request given single body param should translate to request with body 1`] = `
Object {
  "body": Object {
    "contents": Array [
      Object {
        "examples": Array [],
        "id": Any<String>,
        "mediaType": "*",
      },
    ],
    "id": Any<String>,
    "name": "param",
  },
  "cookie": Array [],
  "headers": Array [],
  "path": Array [],
  "query": Array [],
}
`;

exports[`request given single form param should translate to request with form 1`] = `
Object {
  "body": Object {
    "contents": Array [
      Object {
        "id": Any<String>,
        "mediaType": "*",
        "schema": Object {
          "$schema": "http://json-schema.org/draft-07/schema#",
          "properties": Object {
            "param": Object {
              "type": "number",
            },
          },
          "type": "object",
          "x-stoplight": Object {
            "id": Any<String>,
          },
        },
      },
    ],
    "id": Any<String>,
  },
  "cookie": Array [],
  "headers": Array [],
  "path": Array [],
  "query": Array [],
}
`;

exports[`request given single header param should translate to request with header 1`] = `
Object {
  "cookie": Array [],
  "headers": Array [
    Object {
      "id": Any<String>,
      "name": "param",
      "style": "simple",
    },
  ],
  "path": Array [],
  "query": Array [],
}
`;

exports[`request given single path param should translate to request with path 1`] = `
Object {
  "cookie": Array [],
  "headers": Array [],
  "path": Array [
    Object {
      "id": Any<String>,
      "name": "param",
      "required": true,
      "style": "simple",
    },
  ],
  "query": Array [],
}
`;

exports[`request given single query param should translate to request with query 1`] = `
Object {
  "cookie": Array [],
  "headers": Array [],
  "path": Array [],
  "query": Array [
    Object {
      "id": Any<String>,
      "name": "param",
      "style": "unspecified",
    },
  ],
}
`;

exports[`request given two header params should translate 1`] = `
Object {
  "cookie": Array [],
  "headers": Array [
    Object {
      "id": Any<String>,
      "name": "param",
      "style": "simple",
    },
  ],
  "path": Array [],
  "query": Array [],
}
`;

exports[`request given two path params should translate 1`] = `
Object {
  "cookie": Array [],
  "headers": Array [],
  "path": Array [
    Object {
      "id": Any<String>,
      "name": "param",
      "required": true,
      "style": "simple",
    },
  ],
  "query": Array [],
}
`;

exports[`request given two query params should translate 1`] = `
Object {
  "cookie": Array [],
  "headers": Array [],
  "path": Array [],
  "query": Array [
    Object {
      "id": Any<String>,
      "name": "param",
      "style": "unspecified",
    },
  ],
}
`;

exports[`request should translate mixed request 1`] = `
Object {
  "body": Object {
    "contents": Array [
      Object {
        "examples": Array [],
        "id": Any<String>,
        "mediaType": "*",
      },
    ],
    "id": Any<String>,
    "name": "param",
  },
  "cookie": Array [],
  "headers": Array [
    Object {
      "id": Any<String>,
      "name": "param",
      "style": "simple",
    },
  ],
  "path": Array [
    Object {
      "id": Any<String>,
      "name": "param",
      "required": true,
      "style": "simple",
    },
  ],
  "query": Array [
    Object {
      "id": Any<String>,
      "name": "param",
      "style": "unspecified",
    },
  ],
}
`;
